### PR TITLE
Flaml: fix lgbm reproducibility

### LINKED ISSUE #1368: [Bug]: LGBM Results Not Reproducible
### Describe the bug

Currently there are a couple of issues around the reproducibility of the FLAML best loss, when the best estimator found was an LGBMEstimator. 
These issues occur as follows: 

1.  The best loss returned by FLAML is not reproducible, using the underlying LGBMClassifier or LGBMEstimator model (i.e. `automl.model.model`). This seems to be caused by `n_estimators` always being set to 1, regardless of what value it should be. Please note that `n_estimators` always seems to be set correctly on the FLAMLised LGBMEstimator - this issue exists exclusively on the underlying models.
2. With certain configurations, the FLAML best loss can't be reproduced even when using the FLAMLised `LGBMEstimator`. This seems to only be the case when a time budget is set, so the issue is likely caused by the callbacks, similar to the issue found with CatBoostEstimators [here](https://github.com/microsoft/FLAML/pull/1364)

### Steps to reproduce

In terms of showcasing these issues in a reproducible manner:
1. Can be replicated by uncommenting and then running coverage `run -m pytest -k test_reproducibility_of_underlying_regression_models[lgbm]` (or the classification equivalent.
2. A unit test will be added for this, which will initially fail, until the issue is fixed.

I'll open a PR for this shortly - hope that's okay

### Model Used

LGBMEstimators

### Expected Behavior

Taking the FLAMLised LGBMEstimator model, or the underlying model it wraps (LGBMClassifier/LGBMRegressor) and training and testing it on the same folds, should return the same best loss as FLAML.

### Screenshots and logs

_No response_

### Additional Information

_No response_

### PR DESCRIPTION
**Pull request by dannycg1996**
## Why are these changes needed?

In some cases, FLAML LGBM results are not reproducible, using the best model provided by FLAML. I can split the problem, and the solution, into two:

# 1. n_estimators on automl.model.model always = 1

As described in the title, the code `self._model.set_params(n_estimators=best_iteration + 1)` on  `LGBMEstimator.fit()` causes the the `n_estimators` value to always be 1. I'm not sure why `self._model.best_iteration_` is always 0 - I think it's due to using EarlyStopping? Regardless, we have no need to overwrite a correct `n_estimators` value, so I've removed this code (and my unit tests now pass!)

# 2. When a time_budget is set, LGBMEstimator sometimes is not reproducible, if the deadline is running out when the estimator is trained

This is a very similar issue to the one we recently resolved for the CatBoost model. Basically the callback causes the training of a model to differ, when the time budget is running out - thus making the loss impossible to reproduce. I've fixed this by simply deleting the if statement responsible - I think this is a clean solution.

Like with the CatBoost fix, I've also added a unit test, which fails when the if statement above is re-added to the codebase. This will help us identify if the issue is ever reimplemented somehow. It also allows others to test this issue, should they wish.

Also, am I okay to open an issue sometime to refactor this LGBMEstimator? There's a lot of defunct code, and it could be made both more readable and possibly more efficient. 

Anyway, please let me know what you think!

## Related issue number

Closes #1368 

## Checks

- [X] I've used [pre-commit](https://microsoft.github.io/FLAML/docs/Contribute#pre-commit) to lint the changes in this PR (note the same in integrated in our CI checks).
- [X] I've included any doc changes needed for https://microsoft.github.io/FLAML/. See https://microsoft.github.io/FLAML/docs/Contribute#documentation to build and test documentation locally.
- [X] I've added tests (if relevant) corresponding to the changes introduced in this PR.
- [ ] I've made sure all auto checks have passed.

### CODE REVIEW

File: flaml/automl/model.py
Comment (thinkall) [inline flaml/automl/model.py:1592]: This helps control the time budget. In the case of unreproducibility, did you observe that the post-training loss is lower than the training loss?

**Review by thinkall — COMMENTED** [review]
Thank you so much for the PR, @dannycg1996 !

File: flaml/automl/model.py
Comment (dannycg1996) [reply flaml/automl/model.py:1592]: Hi @thinkall, I've tried investigating this issue further, and I'm a little confused. Apologies for the wall of text.

Firstly, in response to this comment, I re-added the if statement in question back to my local code, and reran `test_reproducibility_of_lgbm_regression_model` . My first result was as follows:

image

The reproduced loss is lower than the original loss. In this case, we'd be best off letting the LGBM run to completion.

I then started looking into this early stopping a little more. I tracked when the callback was called, and also when the above early stopping was called. Re-running `test_reproducibility_of_lgbm_regression_model` revealed the following:

- 10 LGBM estimators were tried and tested in total
- The early stopping was called 9 times in total, once for each fold on the very last model (that makes sense).
- **The callback called was 36 times, per model - why is it being called 4 times per fold?**
- **The deadline changes each time a new model is tried - shouldn't there be a single deadline for all models, dictated by the time budget?** I imagine I'm missing something obvious on these last two points.

In summary, I see the trade-offs as being along these lines:

**In favour of Early Stopping**

1. Time budget is respected more.
2. With early stopping, we _maybe_ provide a minimum loss for that model on that fold, and we can potentially surpass that loss if we retrain the model on the same folds, with no time budget/risk of early stopping (I'm not certain this is guaranteed).

**Against Early Stopping**

1. In some cases, results may not be reproducible.
2. In rare cases, the performance of the last model/s may be understated due to early stopping, which could result in user's being provided a less performant model through `automl.model`.

**Things to Note** 

- This early stopping will only ever affect the very last model (or group of models if using BlendSearch). I don't know how much time this may add onto the AutoML runtime, but I have to assume not very much.
- I've had to search for a specific configuration which triggers this issue - the chances of it occurring otherwise are low, albeit still possible.

Thanks for any help - would appreciate your opinions on the above (I know it's a lot). I'd still very much argue in favour of removing this early stopping behaviour, but can accept there might be a few good arguments for retaining it.

Comment (thinkall) [reply flaml/automl/model.py:1592]: Each time `_callback` is called, `EarlyStopException` will be called only when deadline is reached, that's why you see more calls on `_callback`. It's not necessary 4 times. If you change your parameters, you'll see different results.

image
 
Deadline changes because `start_time` is updated at each `fit`.
image

I think we should keep the time budget limitation here, otherwise, it could happen that the run exceeds the time budget too far. It'll happen a lot when the dataset is huge. In most cases, the last configuration wouldn't be picked up as the best parameters. Even it's picked, the retraining on it will get a better performance instead of worse one.

**Comment by dannycg1996** [comment]
Hi @thinkall, thanks for taking the time to explain - I understand why you wish to keep the early stopping!
I've added the early stopping back in as you requested - once tests pass, please feel free to merge.

File: flaml/automl/model.py
Comment (dannycg1996) [reply flaml/automl/model.py:1576]: Hi @thinkall, no worries - best to be cautious here. 

The code highlighted above doesn't overwrite the `LGBMEstimator` `n_estimators` value - it overwrites the underlying `LGBMClassifier` or `LGBMRegressor` value. Please note that this code is also run **after** the `._fit()` has been called, so the highlighted code doesn't actually affect the actual fitting of the model, it just affects the configuration of `automl.model.model` i.e. the underlying model returned to the user.

If I run `test_reproducibility_of_underlying_classification_models[lgbm]` (which is commented out on FLAML main), I get the below error, showing that the FLAML best loss cannot be reproduced using the underlying `LGBMClassifier` model. It _is_ however reproducible using the FLAMLised LGBMEstimator.
image

Please note that for the above screenshot, I'm running this test on the FLAML main branch, with the only difference being that I uncommented this test out (plus a few print statements).

Whilst you are accurate to say that `automl.best_config` is correct, that's not the issue here. `automl.best_config` returns the configuration of the best FLAMLised estimator, in this case, the `LGBMEstimator` configuration. Specifically in this failing test, I am trying to reproduce the FLAML best loss using the underlying `LGBMClassifier` or `LGBMRegressor` (not the FLAMLised estimator - that's fine). 

By investigating the configuration of these underlying models, I can see that every underlying `LGBMClassifier` or `LGBMRegressor` returned by FLAML has `n_estimators=1`. This is because `self._best_iteration_` always returns 0 for these estimators. The configuration of the underlying model is being overwritten _after_ it has been fitted, which doesn't make sense. I can't see why the `n_estimators` values should differ between the `LGBMEstimator` and underlying `LGBMClassifier`.   

If instead of logging the LGBMEstimator configs, we logged the LGBMClassifier configs, you would see that `n_estimators` is always 1 - regardless of the `n_estimators` value which was actually used. In fact, I tried exactly that! I added print statements to the above code i.e.:

```
            best_iteration = (
                getattr(self._model.get_booster(), "best_iteration", None)
                if isinstance(self, XGBoostSklearnEstimator)
                else self._model.best_iteration_
            )
            print('previous n_estimators value', self._model.get_params()['n_estimators'])
            print('best_iteration', best_iteration)
            if best_iteration is not None:
                self._model.set_params(n_estimators=best_iteration + 1)
```

and the following was printed to my console:
image

`best_iteration` is **always** 0, meaning that we **always"" assign the underlying model `n_estimators=1`

I am aware that other estimators depends on this. I've tested XGBoost specifically, and in every test I ran, `best_estimators` was identical to the pre-existing `n_estimators` value. This means that the above line overwrites the `n_estimators` value - with the exact same value. Removing that if statement had no effect on reproducibility.

I think that any errors in other errors (such as XGBoost and RandomForest) which are caused by deleting the above code should cause my already-written reproducibility unit tests failing. However, none currently fail, so I don't think I've caused any errors. However, if you think I've missed something, or have a better suggestion for how to fix this bug, please let me know!

Comment (thinkall) [reply flaml/automl/model.py:1576]: Thank you so much for all the amazing details, @dannycg1996! `best_iteration_` only works when early_stopping is enabled. https://lightgbm.readthedocs.io/en/latest/pythonapi/lightgbm.LGBMClassifier.html#lightgbm.LGBMClassifier.best_iteration_ 

As I understand, best_iteration_  should default to n_estimator instead of 0. I'm not sure why it's 0. Looks like early_stopping is not working correctly here. Without early_stopping, we should not set n_estimators to best_iteration+1. 

What about updating the if statement to `if best_iteration is not None and best_iteration>1`?

Comment (dannycg1996) [reply flaml/automl/model.py]: No worries!

I went for `if best_iteration is not None and best_iteration > 0` in the end - I think it's fine. I did consider simplifying this to just `if best_iteration` as both None and 0 are falsy, but looking at the LGBM codebase, in some places `best_iteration=-1` when initialised, and that would cause a bug!
I'll admit that I'm sceptical about whether this whole `best_iteration` code is actually serving much of a useful purpose for any estimators, but I understand your hesitation in making major tweaks that could have unintended consequences.

I think the issue with best_iteration being returned as 0 might be similar to [this](https://stackoverflow.com/questions/75867049/lgbms-best-iteration-is-none-when-using-early-early-stopping-callback-even-tho) StackOverflow post by the way. Basically with certain settings the LGBM best_iteration_ parameter doesn't work as it should. If I can make an easily reproducible example of this, I may post an issue on the LGBM GitHub at some point.

I also added another (minor change) to the codebase here, as it's been annoying me since I started working on LGBMs! Basically I've merged two if statements i.e. from this:

image

to this:
image

I hope that's OK - it seems a sensible minor refactor to me!

Let me know what you think - thanks!

**Review by thinkall — APPROVED** [review]
Thank you sooooooooo much, @dannycg1996 ! LGTM!

**Comment by dannycg1996** [comment]
Ah, it seems I have an issue with this test which I added:
image

I can look more into this tomorrow, but it seems like it's caused by the EarlyStopping that FLAML implements for LGBMs. We've already discussed this of course, but the one caveat worth mentioning here is that the reproduced score is worse than the FLAML score. That _is_ a problem I think. Please don't merge anything until we've discussed!

**Comment by dannycg1996** [comment]
Sorry about that @thinkall, I think it should be approvable and mergeable now!

**Pull request merged**
